### PR TITLE
Allow to take and then check snapshot status when using pvc

### PR DESCRIPTION
**Pull request by sagor999**
## Description

First part of allowing to take a workspace snapshot that uses PVC.
Second part to make it actually work is for webapp to implement necessary changes.

## Related Issue(s)

Related https://github.com/gitpod-io/gitpod/issues/14159

## Release Notes

```release-note
none
```

## Werft options:

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [x] /werft with-large-vm
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`

### CODE REVIEW

**Comment by roboquat** [comment]
@sagor999: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by jenting** [comment]
Can we use `gp snapshot` to do the test?
Or do we need the WebApp change to test against this PR?

**Comment by sagor999** [comment]
@jenting good idea. I think we can test snapshot creation for sure, not sure if restore would work out of the box or would require some tweaking on webapp side. 
will test it tomorrow.

**Comment by sagor999** [comment]
@jenting `gp snapshot` works by calling takeSnapshot on the server (webapp). Which makes sense, so that it follows same code path. But that also means that until webapp implements their part, we cannot directly call TakeSnapshot on wsman via `gp snapshot`.

File: components/ws-manager/pkg/manager/manager_ee.go [hunk 1]
```diff
@@ -54,21 +56,66 @@ func (m *Manager) TakeSnapshot(ctx context.Context, req *api.TakeSnapshotRequest
...
+		wsClassName := wso.Pod.Labels[workspaceClassLabel]
+
+		workspaceClass := m.Config.WorkspaceClasses[wsClassName]
+		if workspaceClass != nil {
+			pvcVolumeSnapshotClassName = workspaceClass.PVC.SnapshotClass
+		}
+	}
+
+	if pvcFeatureEnabled {
+		workspaceID, _ := wso.WorkspaceID()
```
Comment (jenting) [inline components/ws-manager/pkg/manager/manager_ee.go:72]: Should we check the workspace ID exists before using it?
For example
```go
id, ok := wso.WorkspaceID()
if !ok {
    ....
}
```

Comment (sagor999) [reply components/ws-manager/pkg/manager/manager_ee.go:72]: `sts, err := m.getWorkspaceStatus(*wso)` on line 50 above checks for that already, so not necessary to do it again.

File: components/ws-manager-api/core.proto [hunk 2]
```diff
@@ -44,6 +44,9 @@ service WorkspaceManager {
     // deleteVolumeSnapshot asks ws-manager to delete specific volume snapshot and delete source from cloud provider as well
     rpc DeleteVolumeSnapshot(DeleteVolumeSnapshotRequest) returns (DeleteVolumeSnapshotResponse) {}
 
+    // checkVolumeSnapshotReady checks if a volume snapshot (PVC) is ready
+    rpc CheckVolumeSnapshotReady(CheckVolumeSnapshotReadyRequest) returns (CheckVolumeSnapshotReadyResponse) {}
```
[the remark this answers is not in the document]
Comment (sagor999) [reply components/ws-manager-api/core.proto]: fixed

**Comment by kylos101** [comment]
@sagor999 @jenting :wave: a couple other options to consider for testing:
1. maybe `gpctl workspaces snapshot` will do?
2. if not, can we test the endpoint using https://github.com/fullstorydev/grpcurl?

**Review by jenting — APPROVED** [review]
Code change LGTM

But I need to find a way to verify it so put a hold label

/hold

File: components/ws-manager-api/core.proto [hunk 3]
```diff
@@ -44,6 +44,9 @@ service WorkspaceManager {
     // deleteVolumeSnapshot asks ws-manager to delete specific volume snapshot and delete source from cloud provider as well
     rpc DeleteVolumeSnapshot(DeleteVolumeSnapshotRequest) returns (DeleteVolumeSnapshotResponse) {}
 
+    // CheckVolumeSnapshotIsReady checks if a volume snapshot (PVC) is ready
+    rpc CheckVolumeSnapshotIsReady(CheckVolumeSnapshotIsReadyRequest) returns (CheckVolumeSnapshotIsReadyResponse) {}
```
[the remark this answers is not in the document]
Comment (easyCZ) [reply components/ws-manager-api/core.proto:48]: And it's perfectly fine if for now the only property present on the VolumeSnapshot are `id` and `status` which has a value of "ready"

[the remark this answers is not in the document]
Comment (sagor999) [reply components/ws-manager-api/core.proto]: @easyCZ done. Should be easier to extend now in the future if we have to. PTAL.

**Review by easyCZ — APPROVED** [review]

**Comment by sagor999** [comment]
Pinged @svenefftinge in slack with info. VolumeSnapshot is not passed into StartWorkspace call when restoring from snapshot. Asked for his help to look from webapp side why so.

**Comment by svenefftinge** [comment]
> Pinged @svenefftinge in slack with info. VolumeSnapshot is not passed into StartWorkspace call when restoring from snapshot. Asked for his help to look from webapp side why so.

Maybe I'm confused by snapshot names, ids, handles, etc. but I think we need to return the `snapshotHandle` from ws-amanger's `TakeSnapShot` and `GetVolumeSnapshot`. Looks like it is only set on finalize workspace and then passed through the status update atm.

**Comment by easyCZ** [comment]
Moving this to draft as it:
1. Needs a rebase
2. Hasn't seen any activity in the past couple days

**Comment by sagor999** [comment]
Thank you @easyCZ
Yes, this PR is paused for now as we are re-thinking our PVC approach a bit.

**Pull request closed without merging**